### ISSUE TITLE
Client slowdown and memory leak

### ISSUE THREAD

**Issue by lhofhansl** [minetest/minetest#10499]
Latest MT built from master.

This happens on Linux but seemingly *not* on a MacBook.

Problem:
Fly around in a plain world, notice that the client uses more and more memory. The longer you look around the more memory the client will consume. I stopped close to 6GB.
Also - this maybe related or not - the framerate goes down after a while (down to 10-20 FPS from 60 FPS). When exiting the game (but not the client app) memory consumption is *not* going down, but it's back to 60 FPS.

See screenshots: (16 FPS, time spent on rendering the solid meshes: close to 40ms)
![screenshot_20201013_225356](https://user-images.githubusercontent.com/446130/95949569-c920c680-0da7-11eb-91b1-51148791fcbb.png)

Exited and reentered the game. (The very same scene... back to 60FPS, solid meshes take about 10ms only.)
![screenshot_20201013_225437](https://user-images.githubusercontent.com/446130/95949574-ca51f380-0da7-11eb-9ceb-8122772632dd.png)

Has anybody else seen this? Somehow rendering slows down. And there's a memory leak (may or may not be related)

**Comment by lhofhansl** [comment]
I had suspected `map::m_sectors` (only ever set in `clientmap`), since we never delete from it.
While this does look like a problem, even if I delete sectors from `m_sectors` that out of range, the memory leak continues.

**Comment by benrob0329** [comment]
Can confirm memory leak on Linux, though I have not noticed a significant slowdown because of it (that could be down to horsepower differences though).

**Comment by paramat** [comment]
Possibly a duplicate of #8977 ?

**Comment by lhofhansl** [comment]
Perhaps. Nobody mentions the rendering slowdown. It's actually pretty extreme... It start with 60 FPS, then after just flying forward in a vanilla world for a few minutes FPS reduce to as low as 10 FPS. Exiting the world and reentering gets it back to 60 FPS (but memory is still not freed).

Did some simple memory scanning, and it is indeed the heap that grows without bounds. And since it slows down the actual rendering I suspect the texture cache or something in Irrlicht.

I have relatively little time ATM, if someone would have time to do a valgrind session that'd be quite helpful.

**Comment by paramat** [comment]
The FPS issue might be a duplicate of your issue #7904 ?

**Comment by lhofhansl** [comment]
@paramat Yes it is. I forgot about that. Still no solution :(

**Comment by lhofhansl** [comment]
Hah. Found the problem. Take a look at this code in Irrlicht: https://sourceforge.net/p/irrlicht/code/HEAD/tree/branches/releases/1.8/source/Irrlicht/CNullDriver.cpp#l1556

Notice how hw buffers are only removed after called this 20000 times. updateAllHardwareBuffers is called from driver::endScene, which we do call. But with 60 FPS this cleans out the hw buffers every 20000/60/60 = 5.5 minutes, if rendering slows down much longer.
Calling `driver::removeAllHardwareBuffers()` after each scene is also not good, but if I call it - as a test - every 100's frame the slowdown goes away. *Not the memory leak*, though, so that's an unrelated problem.

**Comment by lhofhansl** [comment]
I also found the "memory leak", by default the client keep up to 5000 blocks loaded from the server for 10 minutes.

And this was related the the problem here, it is the hw-buffers of these block still loaded into the client that causes the severe slowdown.

I'm trying to play with smaller settings for these two now.

At the very least I could try to release the hw buffer for a block when its refcount goes to 0.

**Comment by lhofhansl** [comment]
Ok... I had `client_mapblock_limit` set to -1 (unlimited). D'oh.
So it would accumulate a lot of blocks, and hence a lot of hw buffers.

When I set this back to the default there is no memory leak it seems.
I still see the slowdown due to the hw buffers, but less so (FPS from 60 to 40, not to 20 or less as before).

Still not 100% sure how to work around the broken release of HW buffers in Irrlicht. Ref counts of blocks go to 0 zero too often and we'd prematurely release the HW buffers (I tried that).

I could invent another mechanism in `map::timerUpdate`, which is responsible for releasing the blocks. There would be a second (much shorter) timeout after which we'd release a block meshes' HW buffers.

Or I could do the simple thing I suggested above: After each 1000's frame (i.e. about 17s @ 60 FPS), release all HW buffers. That is simple and pretty safe and I did not see any bad effects as only the required HW buffers are reinitialized in the next frame.

Perhaps I'll find some time tomorrow. As is, it is not as bad as I thought. Although loosing about 1/3 of your FPS is not a laughing matter.
